generic_server.cc

```diff
@@ -229,19 +348,42 @@ server::listen(socket_address addr, std::shared_ptr<seastar::tls::credentials_bu
 }
 
 future<> server::do_accepts(int which, bool keepalive, socket_address server_addr) {
-    return repeat([this, which, keepalive, server_addr] {
+    for (;;) {
         seastar::gate::holder holder(_gate);
-        return _listeners[which].accept().then_wrapped([this, keepalive, server_addr, holder = std::move(holder)] (future<accept_result> f_cs_sa) mutable {
+        bool shed = false;
+        try {
+            semaphore_units<named_semaphore_exception_factory> units(_conns_cpu_concurrency_semaphore, 0);
+            if (_conns_cpu_concurrency != std::numeric_limits<uint32_t>::max()) {
+                auto u = try_get_units(_conns_cpu_concurrency_semaphore, 1);
+                if (u) {
+                    units = std::move(*u);
+                } else {
+                    _blocked_connections++;
+                    try {
+                        units = co_await get_units(_conns_cpu_concurrency_semaphore, 1, std::chrono::minutes(1));
+                    } catch (const semaphore_timed_out&) {
+                        shed = true;
+                    }
+                }
+            }
+            accept_result cs_sa = co_await _listeners[which].accept();
             if (_gate.is_closed()) {
-                f_cs_sa.ignore_ready_future();
-                return stop_iteration::yes;
+                break;
             }
-            auto cs_sa = f_cs_sa.get();
             auto fd = std::move(cs_sa.connection);
             auto addr = std::move(cs_sa.remote_address);
             fd.set_nodelay(true);
             fd.set_keepalive(keepalive);
-            auto conn = make_connection(server_addr, std::move(fd), std::move(addr));
+            auto conn = make_connection(server_addr, std::move(fd), std::move(addr),
+                    _conns_cpu_concurrency_semaphore, std::move(units));
```

Comment:
In the `shed` case, we move zero units to the new connection.
this will leak units as `invoke_with_counting` calls return_all (with 0 units) and then unconditionally call `consume_units` for a single unit even if it returned no units.